### PR TITLE
Add FirstEnabledTraitOrDefault helper method.

### PR DESCRIPTION
**Pull request by RoosterDragon**
This avoids the allocations caused by LINQ when using traits.FirstOrDefault(Exts.IsTraitEnabled). This is important in FrozenActorLayer.RefreshState which is called very often. We apply the new helper method to all areas using the old pattern. An overload that takes an array allows arrays to be enumerated without causing allocations.

### CODE REVIEW

File: OpenRA.Mods.Common/Lint/CheckTooltips.cs
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Lint/CheckTooltips.cs]: By the way, does `Exts.IsTraitEnabled` work on `TraitInfo`s? I figured it was just for traits (hoping to learn something :))

Comment (pchote) [reply OpenRA.Mods.Common/Lint/CheckTooltips.cs]: It doesn't. This is a copy-pasta error of an older logical error.  Lets get rid of it.

Comment (pchote) [reply OpenRA.Mods.Common/Lint/CheckTooltips.cs]: You can use `ConditionalTraitInfo.EnabledByDefault` instead.

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Lint/CheckTooltips.cs]: A cursory glance shows places other than just tooltips affected, I have created #14445 to track this instead.

**Comment by reaperrr** [comment]
@RoosterDragon I suspect we have quite a few cases of `traits.FirstOrDefault(t => !t.IsTraitDisabled)` as well, would it win us anything to make those use this helper method, too?

File: OpenRA.Mods.Common/Lint/CheckTooltips.cs
Comment (reaperrr) [reply OpenRA.Mods.Common/Lint/CheckTooltips.cs]: Wouldn't it still be better to fix the places that are touched by this PR, though?

The old bogus look-up would just ignore the Exts.IsTraitsEnabled, but now this uses trait-specific look-up for TraitInfo, will this even work (without causing regressions vs. bleed, I mean)?

Besides, I couldn't find any other places (or rather, the only other similar place I could find already uses EnabledByDefault as @pchote suggested), though it's possible my search approach didn't catch them all.

**Comment by RoosterDragon** [comment]
Have fixed bad `Exts.IsTraitEnabled` calls affected by this PR. Remaining bad uses are covered in #14445.

**Review by reaperrr — APPROVED** [review]

File: OpenRA.Game/Exts.cs
Comment (pchote) [inline OpenRA.Game/Exts.cs]: General question: Is the compiler smart enough to generate code to enumerate these by index (`for (var i = 0; i < ts.Length; i++)`, or does it allocate an enumerator object?

**Review by pchote — APPROVED** [review]
LGTM, but i'm interested in finding out the answer to my question above before hitting the merge button.

File: OpenRA.Game/Exts.cs
Comment (pchote) [reply OpenRA.Game/Exts.cs]: According to the internet: yes.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5d8c76f4d51f7c75fe38bed47b3bf2c70e6493f0

File: OpenRA.Game/Exts.cs
Comment (RoosterDragon) [reply OpenRA.Game/Exts.cs]: Yes, the compiler special cases `foreach` on arrays to do exactly that.

**Pull request merged**
